clippy_lints/src/call_missing_target_feature.rs

```diff
@@ -0,0 +1,141 @@
+    /// #[target_feature(enable = "avx2")]
+    /// unsafe fn f() -> u32 {
+    ///     0
+    /// }
+    ///
+    /// fn g() {
+    ///     unsafe { f() };
+    ///     // g does not enable the target features f requires
+    /// }
+    /// ```
+    #[clippy::version = "1.82.0"]
+    pub CALL_MISSING_TARGET_FEATURE,
+    suspicious,
+    "call requires target features that the surrounding function does not enable"
+}
+
+declare_lint_pass!(CallMissingTargetFeature => [CALL_MISSING_TARGET_FEATURE]);
+
+impl<'tcx> LateLintPass<'tcx> for CallMissingTargetFeature {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &hir::Expr<'tcx>) {
+        let Some(callee_def_id) = callee_def_id(cx, expr) else {
+            return;
+        };
+        let callee_target_features = &cx.tcx.codegen_fn_attrs(callee_def_id).target_features;
+
+        if callee_target_features.is_empty() {
+            return;
+        }
+
+        let Some(caller_body_id) = cx.enclosing_body else {
+            return;
+        };
+        let caller_def_id = cx.tcx.hir().body_owner_def_id(caller_body_id);
+        let caller_target_features = &cx.tcx.codegen_fn_attrs(caller_def_id).target_features;
+
+        if in_external_macro(cx.tcx.sess, expr.span) {
+            return;
+        }
+
+        // target features can imply other target features (e.g. avx2 implies sse4.2). We can safely skip
+        // implied target features and only warn for the more general missing target feature.
+        let missing: Vec<_> = callee_target_features
+            .iter()
+            .filter_map(|target_feature| {
+                if target_feature.implied || caller_target_features.iter().any(|tf| tf.name == target_feature.name) {
+                    None
+                } else {
+                    Some(target_feature.name.as_str())
+                }
+            })
+            .collect();
+
+        if missing.is_empty() {
+            return;
+        }
+
+        let attr = format!("#[target_feature(enable = \"{}\")]", missing.join(","));
+
+        span_lint_and_then(
+            cx,
+            CALL_MISSING_TARGET_FEATURE,
+            expr.span,
+            "this call requires target features that the surrounding function does not enable",
+            |diag| {
+                diag.span_label(
+                    expr.span,
+                    "this function call requires target features to be enabled".to_string(),
+                );
+
+                let fn_sig = cx.tcx.fn_sig(caller_def_id).skip_binder();
+
+                let mut suggestions = Vec::with_capacity(2);
+
+                let hir::Node::Item(caller_item) = cx.tcx.hir_node_by_def_id(caller_def_id) else {
+                    return;
+                };
+
+                let Some(indent) = clippy_utils::source::snippet_indent(cx, caller_item.span) else {
+                    return;
+                };
+
+                let lo_span = caller_item.span.with_hi(caller_item.span.lo());
```

Comment:
Use `shrink_to_lo`.

Reply:
Right, I'll fix that along with the category when the FCP is completed